Nullpointer on AbstractHandler#isStateSameAsElementState

The element instance in the context can be null.

```java
java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:169) ~[zb-db-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.reprocessEvent(StreamProcessorController.java:248) ~[zb-logstreams-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.reprocessNextEvent(StreamProcessorController.java:229) ~[zb-logstreams-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:87) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zb-util-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
Caused by: java.lang.NullPointerException
	at io.zeebe.broker.workflow.processor.handlers.AbstractHandler.isStateSameAsElementState(AbstractHandler.java:80) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.handlers.element.ElementActivatingHandler.shouldHandleState(ElementActivatingHandler.java:53) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.handlers.AbstractHandler.handle(AbstractHandler.java:48) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.BpmnStepHandlers.handle(BpmnStepHandlers.java:152) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.BpmnStepProcessor.processRecord(BpmnStepProcessor.java:76) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:56) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedStreamProcessor$DelegatingEventProcessor.processEvent(TypedStreamProcessor.java:179) ~[zeebe-broker-core-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.runInNewTransaction(ZeebeTransactionDb.java:177) ~[zb-db-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:166) ~[zb-db-0.17.0-SNAPSHOT.jar:0.17.0-SNAPSHOT]
	... 8 more
```


```java
  protected boolean isStateSameAsElementState(BpmnStepContext<T> context) {
    return context.getState() == context.getElementInstance().getState();
  }
```

--- a/broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/AbstractHandler.java
+++ b/broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/AbstractHandler.java
@@ -77,7 +77,8 @@ public abstract class AbstractHandler<T extends ExecutableFlowElement>
    * ACTIVATED, and we shouldn't process the ELEMENT_ACTIVATING in that case).
    */
   protected boolean isStateSameAsElementState(BpmnStepContext<T> context) {
-    return context.getState() == context.getElementInstance().getState();
+    return context.getElementInstance() != null
+        && context.getState() == context.getElementInstance().getState();
   }
 
   protected boolean isElementActive(ElementInstance instance) {

--- a/broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletedHandler.java
+++ b/broker-core/src/main/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletedHandler.java
@@ -40,7 +40,7 @@ public class ElementCompletedHandler<T extends ExecutableFlowNode>
   protected boolean shouldHandleState(BpmnStepContext<T> context) {
     return super.shouldHandleState(context)
         && isStateSameAsElementState(context)
-        && isElementActive(context.getFlowScopeInstance());
+        && (isRootScope(context) || isElementActive(context.getFlowScopeInstance()));
   }
 
   @Override

--- a/broker-core/src/test/java/io/zeebe/broker/util/ZeebeStateRule.java
+++ b/broker-core/src/test/java/io/zeebe/broker/util/ZeebeStateRule.java
@@ -17,6 +17,7 @@
  */
 package io.zeebe.broker.util;
 
+import io.zeebe.broker.logstreams.processor.KeyGenerator;
 import io.zeebe.broker.logstreams.state.DefaultZeebeDbFactory;
 import io.zeebe.broker.logstreams.state.ZbColumnFamilies;
 import io.zeebe.broker.logstreams.state.ZeebeState;
@@ -59,6 +60,10 @@ public class ZeebeStateRule extends ExternalResource {
 
   public ZeebeState getZeebeState() {
     return zeebeState;
+  }
+
+  public KeyGenerator getKeyGenerator() {
+    return zeebeState.getKeyGenerator();
   }
 
   public ZeebeDb<ZbColumnFamilies> createNewDb() {

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementActivatedHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementActivatedHandlerTest.java
@@ -1,0 +1,81 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementActivatedHandlerTest extends ElementHandlerTestCase {
+  private final WorkflowInstanceIntent nextState = WorkflowInstanceIntent.ELEMENT_COMPLETING;
+  private ElementActivatedHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementActivatedHandler<>(nextState);
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfFlowScopeIsNotActivated() {
+    // given
+    final ElementInstance flowScope =
+        newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED, flowScope);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldHandleStateIfElementIsRootFlowScope() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isTrue();
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementActivatingHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementActivatingHandlerTest.java
@@ -1,0 +1,121 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.processor.handlers.IOMappingHelper;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.msgpack.mapping.MappingException;
+import io.zeebe.protocol.ErrorType;
+import io.zeebe.protocol.impl.record.value.incident.IncidentRecord;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import io.zeebe.util.buffer.BufferUtil;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementActivatingHandlerTest extends ElementHandlerTestCase {
+  @Mock public IOMappingHelper ioMappingHelper;
+
+  private final WorkflowInstanceIntent nextState = WorkflowInstanceIntent.ELEMENT_ACTIVATED;
+  private ElementActivatingHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementActivatingHandler<>(nextState, ioMappingHelper);
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfFlowScopeIsNotActivated() {
+    // given
+    final ElementInstance flowScope =
+        newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING, flowScope);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldHandleStateIfElementIsRootFlowScope() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isTrue();
+  }
+
+  @Test
+  public void shouldApplyInputMappings() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+
+    // when
+    handler.handle(context);
+
+    // then
+    verify(ioMappingHelper, times(1)).applyInputMappings(context);
+  }
+
+  @Test
+  public void shouldRaiseIncidentOnMappingException() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+
+    // when
+    doThrow(new MappingException("fail")).when(ioMappingHelper).applyInputMappings(context);
+    final boolean handled = handler.handleState(context);
+
+    // then
+    final IncidentRecord raisedIncident = getRaisedIncident();
+    assertThat(handled).isFalse();
+    assertThat(raisedIncident)
+        .extracting(IncidentRecord::getErrorType, IncidentRecord::getErrorMessage)
+        .containsExactly(ErrorType.IO_MAPPING_ERROR, BufferUtil.wrapString("fail"));
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletedHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletedHandlerTest.java
@@ -1,0 +1,158 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.broker.workflow.state.StoredRecord.Purpose;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceRecord;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementCompletedHandlerTest extends ElementHandlerTestCase {
+  private ElementCompletedHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementCompletedHandler<>();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfFlowScopeIsNotActivated() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED, flowScope);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldHandleStateIfElementIsRootFlowScope() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isTrue();
+  }
+
+  @Test
+  public void shouldPublishDeferredRecords() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    final int currentTokens = flowScope.getNumberOfActiveTokens();
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED, flowScope);
+    final ElementInstance deferred = newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+    context
+        .getElementInstanceState()
+        .storeRecord(
+            deferred.getKey(),
+            instance.getKey(),
+            deferred.getValue(),
+            deferred.getState(),
+            Purpose.DEFERRED);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    assertThat(flowScope.getNumberOfActiveTokens()).isEqualTo(currentTokens + 1);
+    verify(eventOutput, times(1))
+        .appendFollowUpEvent(deferred.getKey(), deferred.getState(), deferred.getValue());
+  }
+
+  @Test
+  public void shouldCompleteFlowScopeIfLastActiveElement() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED, flowScope);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, times(1))
+        .appendFollowUpEvent(
+            flowScope.getKey(), WorkflowInstanceIntent.ELEMENT_COMPLETING, flowScope.getValue());
+  }
+
+  @Test
+  public void shouldNotCompleteFlowScopeIfNonePresent() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, never())
+        .appendFollowUpEvent(
+            anyLong(), any(WorkflowInstanceIntent.class), any(WorkflowInstanceRecord.class));
+  }
+
+  @Test
+  public void shouldNotCompleteFlowScopeIfMoreActiveElementsPresent() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETED, flowScope);
+    newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED, flowScope);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, never())
+        .appendFollowUpEvent(
+            eq(flowScope.getKey()), any(WorkflowInstanceIntent.class), eq(flowScope.getValue()));
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletingHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementCompletingHandlerTest.java
@@ -1,0 +1,126 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.processor.handlers.IOMappingHelper;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.msgpack.mapping.MappingException;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementCompletingHandlerTest extends ElementHandlerTestCase {
+  @Mock public IOMappingHelper ioMappingHelper;
+
+  private ElementCompletingHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementCompletingHandler<>(ioMappingHelper);
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfFlowScopeIsNotActivated() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING, flowScope);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldHandleStateIfElementIsRootFlowScope() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isTrue();
+  }
+
+  @Test
+  public void shouldApplyOutputMappings() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(ioMappingHelper, times(1)).applyOutputMappings(context);
+  }
+
+  @Test
+  public void shouldRaiseIncidentOnMappingException() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+    final MappingException exception = new MappingException("failed");
+
+    // when
+    doThrow(exception).when(ioMappingHelper).applyOutputMappings(context);
+    final boolean handled = handler.handleState(context);
+
+    // then
+    assertThat(handled).isFalse();
+    verifyIncidentRaised();
+  }
+
+  @Test
+  public void shouldHandleState() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+
+    // when
+    final boolean handled = handler.handleState(context);
+
+    // then
+    assertThat(handled).isTrue();
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementHandlerTestCase.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementHandlerTestCase.java
@@ -1,0 +1,130 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.logstreams.processor.TypedRecord;
+import io.zeebe.broker.logstreams.processor.TypedStreamWriter;
+import io.zeebe.broker.util.MockTypedRecord;
+import io.zeebe.broker.util.ZeebeStateRule;
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.processor.BpmnStepContext;
+import io.zeebe.broker.workflow.processor.CatchEventBehavior;
+import io.zeebe.broker.workflow.processor.EventOutput;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.protocol.clientapi.RecordType;
+import io.zeebe.protocol.clientapi.ValueType;
+import io.zeebe.protocol.impl.record.RecordMetadata;
+import io.zeebe.protocol.impl.record.value.incident.IncidentRecord;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceRecord;
+import io.zeebe.protocol.intent.IncidentIntent;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+
+public abstract class ElementHandlerTestCase {
+  @ClassRule public static ZeebeStateRule zeebeStateRule = new ZeebeStateRule();
+
+  @Mock public EventOutput eventOutput;
+  @Mock public CatchEventBehavior catchEventBehavior;
+  @Mock public TypedStreamWriter streamWriter;
+  @Captor public ArgumentCaptor<IncidentRecord> incidentCaptor;
+
+  protected BpmnStepContext<ExecutableFlowNode> context;
+
+  @Before
+  public void setUp() {
+    context =
+        new BpmnStepContext<>(
+            zeebeStateRule.getZeebeState().getWorkflowState(), eventOutput, catchEventBehavior);
+    context.setStreamWriter(streamWriter);
+  }
+
+  protected IncidentRecord getRaisedIncident() {
+    verifyIncidentRaised();
+    return incidentCaptor.getValue();
+  }
+
+  protected void verifyIncidentRaised() {
+    verify(eventOutput, times(1)).storeFailedRecord(context.getRecord());
+    verify(streamWriter, times(1))
+        .appendNewCommand(eq(IncidentIntent.CREATE), incidentCaptor.capture());
+  }
+
+  protected ElementInstance createAndSetContextElementInstance(WorkflowInstanceIntent state) {
+    final ElementInstance instance = newElementInstance(state);
+    setContextElementInstance(instance);
+
+    return instance;
+  }
+
+  protected ElementInstance createAndSetContextElementInstance(
+      WorkflowInstanceIntent state, ElementInstance flowScope) {
+    final ElementInstance instance = newElementInstance(state, flowScope);
+    setContextElementInstance(instance);
+    context.setFlowScopeInstance(flowScope);
+
+    return instance;
+  }
+
+  protected void setContextElementInstance(ElementInstance instance) {
+    context.setElementInstance(instance);
+    context.setRecord(newRecordFor(instance));
+  }
+
+  protected ElementInstance newElementInstance(WorkflowInstanceIntent state) {
+    final long key = zeebeStateRule.getKeyGenerator().nextKey();
+    final WorkflowInstanceRecord value = new WorkflowInstanceRecord();
+
+    return zeebeStateRule
+        .getZeebeState()
+        .getWorkflowState()
+        .getElementInstanceState()
+        .newInstance(key, value, state);
+  }
+
+  protected ElementInstance newElementInstance(
+      WorkflowInstanceIntent state, ElementInstance flowScope) {
+    final long key = zeebeStateRule.getKeyGenerator().nextKey();
+    final WorkflowInstanceRecord value = new WorkflowInstanceRecord();
+    value.setFlowScopeKey(flowScope.getKey());
+    flowScope.spawnToken();
+
+    return zeebeStateRule
+        .getZeebeState()
+        .getWorkflowState()
+        .getElementInstanceState()
+        .newInstance(flowScope, key, value, state);
+  }
+
+  protected TypedRecord<WorkflowInstanceRecord> newRecordFor(ElementInstance instance) {
+    final RecordMetadata metadata =
+        new RecordMetadata()
+            .recordType(RecordType.EVENT)
+            .valueType(ValueType.WORKFLOW_INSTANCE)
+            .intent(instance.getState());
+
+    return new MockTypedRecord<>(instance.getKey(), metadata, instance.getValue());
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementTerminatedHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementTerminatedHandlerTest.java
@@ -1,0 +1,128 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceRecord;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementTerminatedHandlerTest extends ElementHandlerTestCase {
+  private ElementTerminatedHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementTerminatedHandler<>();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldTerminateScope() {
+    // given
+    final ElementInstance flowScope =
+        newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, times(1))
+        .appendFollowUpEvent(
+            flowScope.getKey(), WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope.getValue());
+  }
+
+  @Test
+  public void shouldNotTerminateScopeIfRootScope() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, never())
+        .appendFollowUpEvent(
+            anyLong(), any(WorkflowInstanceIntent.class), any(WorkflowInstanceRecord.class));
+  }
+
+  @Test
+  public void shouldNotTerminateScopeIfScopeActive() {
+    // given
+    final ElementInstance flowScope = newElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, never())
+        .appendFollowUpEvent(
+            flowScope.getKey(), WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope.getValue());
+  }
+
+  @Test
+  public void shouldNotTerminateScopeIfActiveSiblingElements() {
+    // given
+    final ElementInstance flowScope =
+        newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    newElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING, flowScope);
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope);
+
+    // when
+    handler.handleState(context);
+
+    // then
+    verify(eventOutput, never())
+        .appendFollowUpEvent(
+            flowScope.getKey(), WorkflowInstanceIntent.ELEMENT_TERMINATED, flowScope.getValue());
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementTerminatingHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/ElementTerminatingHandlerTest.java
@@ -1,0 +1,92 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.zeebe.broker.incident.processor.IncidentState;
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.protocol.ErrorType;
+import io.zeebe.protocol.impl.record.value.incident.IncidentRecord;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class ElementTerminatingHandlerTest extends ElementHandlerTestCase {
+  private ElementTerminatingHandler<ExecutableFlowNode> handler;
+  private IncidentState incidentState = zeebeStateRule.getZeebeState().getIncidentState();
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new ElementTerminatingHandler<>(incidentState);
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    context.setElementInstance(null);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfInMemoryStateIsDifferent() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    instance.setState(WorkflowInstanceIntent.ELEMENT_COMPLETING);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldResolvePendingIncidents() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING);
+    final long incidentKey = zeebeStateRule.getKeyGenerator().nextKey();
+    final IncidentRecord incident =
+        new IncidentRecord()
+            .setElementInstanceKey(instance.getKey())
+            .setVariableScopeKey(instance.getKey())
+            .setErrorMessage("message")
+            .setErrorType(ErrorType.IO_MAPPING_ERROR)
+            .setBpmnProcessId(instance.getValue().getBpmnProcessId())
+            .setElementId(instance.getValue().getElementId())
+            .setWorkflowInstanceKey(instance.getValue().getWorkflowInstanceKey());
+    incidentState.createIncident(incidentKey, incident);
+
+    // when
+    final boolean handled = handler.handleState(context);
+
+    // then
+    assertThat(handled).isTrue();
+    verify(eventOutput, times(1)).appendResolvedIncidentEvent(eq(incidentKey), eq(incident));
+  }
+}

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/EventOccurredHandlerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/handlers/element/EventOccurredHandlerTest.java
@@ -1,0 +1,105 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor.handlers.element;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.zeebe.broker.workflow.model.element.ExecutableFlowNode;
+import io.zeebe.broker.workflow.state.ElementInstance;
+import io.zeebe.protocol.intent.WorkflowInstanceIntent;
+import java.util.EnumSet;
+import java.util.Set;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class EventOccurredHandlerTest extends ElementHandlerTestCase {
+  private EventOccurredHandler<ExecutableFlowNode> handler;
+
+  @Override
+  @Before
+  public void setUp() {
+    super.setUp();
+    handler = new EventOccurredHandler<>();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfNoElementGiven() {
+    // given
+    final ElementInstance instance =
+        createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+    instance.getValue().setWorkflowInstanceKey(zeebeStateRule.getKeyGenerator().nextKey());
+    context.setElementInstance(null);
+    zeebeStateRule
+        .getZeebeState()
+        .getWorkflowState()
+        .getElementInstanceState()
+        .removeInstance(instance.getKey());
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isFalse();
+  }
+
+  @Test
+  public void shouldNotHandleStateIfElementIsNotActive() {
+    // given
+    final Set<WorkflowInstanceIntent> inactiveStates =
+        EnumSet.complementOf(EnumSet.of(WorkflowInstanceIntent.ELEMENT_ACTIVATED));
+
+    // when - then
+    for (final WorkflowInstanceIntent inactiveState : inactiveStates) {
+      final ElementInstance instance = createAndSetContextElementInstance(inactiveState);
+      instance.getValue().setWorkflowInstanceKey(zeebeStateRule.getKeyGenerator().nextKey());
+
+      assertThat(handler.shouldHandleState(context)).isFalse();
+    }
+  }
+
+  @Test
+  public void shouldHandleStateIfElementHasNoWorkflowInstance() {
+    // given
+    final Set<WorkflowInstanceIntent> inactiveStates =
+        EnumSet.complementOf(EnumSet.of(WorkflowInstanceIntent.ELEMENT_ACTIVATED));
+
+    // when - then
+    for (final WorkflowInstanceIntent inactiveState : inactiveStates) {
+      createAndSetContextElementInstance(inactiveState);
+
+      assertThat(handler.shouldHandleState(context)).isTrue();
+    }
+  }
+
+  @Test
+  public void shouldHandleStateIfElementIsActive() {
+    // given
+    createAndSetContextElementInstance(WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+
+    // when - then
+    assertThat(handler.shouldHandleState(context)).isTrue();
+  }
+
+  @Override
+  protected ElementInstance createAndSetContextElementInstance(WorkflowInstanceIntent state) {
+    final ElementInstance instance = super.createAndSetContextElementInstance(state);
+    context.getRecord().getMetadata().intent(WorkflowInstanceIntent.EVENT_OCCURRED);
+
+    return instance;
+  }
+}

--- a/msgpack-value/src/main/java/io/zeebe/msgpack/property/BaseProperty.java
+++ b/msgpack-value/src/main/java/io/zeebe/msgpack/property/BaseProperty.java
@@ -129,10 +129,8 @@ public abstract class BaseProperty<T extends BaseValue> implements Recyclable {
     }
 
     final BaseProperty<?> that = (BaseProperty<?>) o;
-    return isSet == that.isSet
-        && Objects.equals(getKey(), that.getKey())
-        && Objects.equals(value, that.value)
-        && Objects.equals(defaultValue, that.defaultValue);
+    return Objects.equals(getKey(), that.getKey())
+        && Objects.equals(resolveValue(), that.resolveValue());
   }
 
   @Override

